File: writing_an_interpreter_in_go/repl/repl.go
package repl

import (
    "bufio"
    "fmt"
    "io"
    "wl/lexer"
    "wl/parser"
    "wl/token"
)

const PROMPT = ">> "

type EvalFunc func(string) string

func Start(in io.Reader, out io.Writer, eval EvalFunc) {
    scanner := bufio.NewScanner(in)

    for {
        fmt.Printf(PROMPT)
        scanned := scanner.Scan()
        if !scanned {
            return
        }

        line := scanner.Text()
        evaluated := eval(line)
        if evaluated != "" {
            fmt.Println(evaluated)
        }
    }
}


func Lexer_eval(input string) string {
    l := lexer.New(input)

    for tok := l.NextToken(); tok.Type != token.EOF; tok = l.NextToken() {
        fmt.Printf("%+v\n", tok)
    }

    return ""
}


func Parser_eval(input string) string {
    l := lexer.New(input)
    p := parser.New(l)

    program := p.ParseProgram()
    if len(p.Errors()) != 0 {
        fmt.Println("Parser has errors:")
        for _, msg := range p.Errors() {
            fmt.Println("\t" + msg)
        }
        return ""
    }

    return program.String()
}

File: writing_an_interpreter_in_rust/src/cli/repl.rs
use std::io::{self, Write};
use crate::lexer::Lexer;
use crate::parser::Parser;

pub fn repl(eval: fn(&str) -> String) {
    loop {
        print!("> ");
        // Flush stdout to ensure the prompt is displayed
        io::stdout().flush().unwrap();

        // Read user input
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");

        // Trim the input to remove any trailing newline characters
        let input = input.trim();

        // Check for exit command
        if input == "exit" || input == "quit" {
            break;
        }

        // Evaluate the input
        let result = eval(input);

        // Print the result
        println!("{}", result);
    }
}

fn lex(input: &str) -> String {

    let mut lexer = Lexer::new(input.to_string());
    let mut tokens = Vec::new();

    loop {
        let token = lexer.next_token();
        tokens.push(token.clone());

        if token.token_type == crate::token::TokenType::EOF {
            break;
        }
    }
    tokens.iter().map(|t| format!("{:?}", t)).collect::<Vec<String>>().join("\n")
}

fn parse(input: &str) -> String {
    let lexer = Lexer::new(input.to_string());
    let mut parser = Parser::new(lexer);
    let program = parser.parse_program();
    if parser.has_errors() {
        let mut s = format!("Parser errors:\n");
        for e in parser.errors().iter() {
            s.push_str(&format!("\t{}\n", e));
        }
        return s;
    }


    format!("{:?}", program)
}

pub fn repl_command() -> seahorse::Command {
    seahorse::Command::new("repl")
        .description("Starts the REPL")
         .flag(
            seahorse::Flag::new("step", seahorse::FlagType::String)
                .description("Which step to run")
                .alias("s"),
        )
        .action(|c| {
            if let Ok(step) = c.string_flag("step") {
                match step.as_str() {
                    "lexer" => repl(lex),
                    "lex" => repl(lex),
                    "parser" => repl(parse),
                    "parse" => repl(parse),
                    _ => println!("Invalid step"),
                }
                return;
            }
            println!("You need to provide a step");
        })
}
